app/controllers/documents_controller.rb

```diff
@@ -4,19 +4,58 @@ class DocumentsController < ApplicationController
 
   respond_to :html, :json, :js
 
+  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
   def create
-    respond_with do |format|
-      format.html do
-        redirect_to edit_steps_evidence_upload_path
-      end
-      format.json do
-        render json: {}, status: :created
+    document = Document.create_from_file(
+      file: file_from_params, bundle: current_document_bundle
+    )
+
+    Datastore::Documents::Upload.new(document:).call if document.valid?(:criteria)
+
+    respond_with(document, location: evidence_upload_step) do |format|
+      if document.invalid?(:storage)
+        format.html { redirect_to evidence_upload_step }
+        format.json do
+          render json: document.as_json.merge(error_message: document.errors.first.full_message),
+                 status: :unprocessable_entity
+        end
       end
     end
   end
+  # rubocop:enable Metrics/AbcSize, Metrics/MethodLength
 
   def download; end
 
   def destroy; end
+
+  private
+
+  def evidence_upload_step
+    edit_steps_evidence_upload_path(current_crime_application)
+  end
+
+  def current_document_bundle
+    @current_document_bundle ||= DocumentBundle.find(document_bundle_id)
+  end
+
+  # TODO: unify if possible the submission params
+  # Currently XHR and HTML requests have different structure
+  # Also needs proper error handling for non-JS version if user click
+  # upload button without having selected a file
```

Comment:
Currently, this routes to the generic there is 'something wrong with the service' page. Can write a ticket to do this but i think this should go in to unblock other tickets